test/distributed/fsdp/test_fsdp_backward_prefetch.py

```diff
@@ -0,0 +1,202 @@
+
+        torch.manual_seed(0)
+        policy = ModuleWrapPolicy(
+            {nn.TransformerEncoderLayer, nn.TransformerDecoderLayer}
+        )
+        model = FSDP(
+            nn.Transformer(d_model=1024, nhead=8, device="cuda"),
+            device_id=torch.cuda.current_device(),
+            auto_wrap_policy=policy,
+            use_orig_params=True,
+            backward_prefetch=backward_prefetch,
+        )
+        optim = torch.optim.SGD(model.parameters(), lr=1e-2)
+
+        # prepare input
+        torch.manual_seed(rank + 1)
+        src = torch.randn((10, 1, 1024), device="cuda")
+        tgt = torch.randn((20, 1, 1024), device="cuda")
+
+        # monkey patch
+        flat_param_fqns_array = []
+
+        def patched_get_handle_to_prefetch(*args, **kwargs):
+            handle = orig_get_handle_to_prefetch(*args, **kwargs)
+
+            self.assertEqual(
+                len(args), 2, "expect _get_handle_to_prefetch(state, current_handle)"
+            )
+            state = args[0]
+            current_handle = args[1]
+            training_state = _get_training_state(current_handle)
+            if (
+                training_state == HandleTrainingState.BACKWARD_PRE
+                and state.backward_prefetch == BackwardPrefetch.BACKWARD_PRE
+            ) or (
+                training_state == HandleTrainingState.BACKWARD_POST
+                and state.backward_prefetch == BackwardPrefetch.BACKWARD_POST
+            ):
+                nonlocal flat_param_fqns_array
+                fqns = get_flat_param_fqns(state, handle)
+                flat_param_fqns_array.append(fqns)
+            return handle
+
+        # flat params from prefetch handle should match
+        # DECODER_PARAM_FQNS and ENCODER_PARAM_FQNS
+        with patch(
+            "torch.distributed.fsdp._runtime_utils._get_handle_to_prefetch",
+            patched_get_handle_to_prefetch,
+        ):
+            for _ in range(NUM_ITERS):
+                optim.zero_grad()
+                loss = model(src, tgt).sum()
+                loss.backward()
+                optim.step()
+                if backward_prefetch is None:
+                    self.assertEqual(len(flat_param_fqns_array), 0)
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_PRE:
+                    encoder_begin_index = ENCODER_BEGIN_INDEX_FOR_PRE
+                    # +1 is for None handle
+                    self.assertEqual(len(flat_param_fqns_array), TOTAL_NUM_PREFETCH + 1)
+                elif backward_prefetch == BackwardPrefetch.BACKWARD_POST:
+                    encoder_begin_index = ENCODER_BEGIN_INDEX_FOR_POST
+                    self.assertEqual(len(flat_param_fqns_array), TOTAL_NUM_PREFETCH + 1)
+
+                for array_index, fqns in enumerate(flat_param_fqns_array):
+                    if array_index >= 0 and array_index < encoder_begin_index:
+                        param_index = encoder_begin_index - 1 - array_index
```

Comment:
`param_index` seems more like `layer_index`? This is the layer index in the `nn.ModuleList`?